src/ec/eddsa.rs

```diff
@@ -176,10 +176,56 @@ impl signature::VerificationAlgorithm for EdDSAParameters {
         }
         let msg = msg.as_slice_less_safe();
         let signature = signature.as_slice_less_safe();
-        bssl::map_result(unsafe {
-            GFp_ed25519_verify(msg.as_ptr(), msg.len(), signature.as_ptr(),
-                               public_key.as_ptr())
-        })
+
+        let mut a = Point::new();
+        if (signature[63] & 224) != 0 {
+            return Err(error::Unspecified);
+        }
+        let ret = unsafe {
+            GFp_x25519_ge_frombytes_vartime(&mut a, public_key.as_ptr())
+        };
+        if ret != 0 { return Err(error::Unspecified) }
+
+        for i in 0 .. ELEM_LIMBS {
+            a.x[i] = -a.x[i];
+            a.t[i] = -a.t[i];
+        }
+        let mut r_copy = [0u8; ELEM_LEN];
+        unsafe {
+            core::ptr::copy_nonoverlapping(signature.as_ptr(),
+                                           r_copy.as_mut_ptr(), ELEM_LEN)
+        };
+        let mut s_copy = [0u8; ELEM_LEN];
+        unsafe {
+            core::ptr::copy_nonoverlapping(signature.as_ptr().offset(32),
+                                           s_copy.as_mut_ptr(), ELEM_LEN)
+        };
+        let mut h = [0u8; digest::SHA512_OUTPUT_LEN];
+        digest::GFp_SHA512_4(h.as_mut_ptr(), h.len(),
+                             signature.as_ptr(), 32,
+                             public_key.as_ptr(), 32,
+                             msg.as_ptr(), msg.len(),
+                             core::ptr::null(), 0);
+        unsafe {
+            GFp_x25519_sc_reduce(h.as_mut_ptr())
+        };
+        let mut r = ProjPoint::new();
+        unsafe {
+            GFp_ge_double_scalarmult_vartime(&mut r, h.as_ptr(), &a,
+                                             (&s_copy).as_ptr());
+        };
+        let mut r_check = [0u8; ELEM_LEN];
+        unsafe {
+            GFp_x25519_ge_tobytes(r_check.as_mut_ptr(), &r)
+        };
+        let cmp = unsafe {
+            constant_time::GFp_memcmp(r_check.as_ptr(), (&r_copy).as_ptr(), 32)
```

Comment:
This should just use the safe `verify_slices_are_equal()` function. Also maybe instead of serializing `r` to `r_check` and then comparing `r_check` to `r_copy`, we should just compare the deserialized values?